File: Python3/1080_insufficient_nodes_in_root_to_leaf_paths.py
# https://leetcode.cn/problems/insufficient-nodes-in-root-to-leaf-paths/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sufficientSubset(self, root: Optional[TreeNode], limit: int) -> Optional[TreeNode]:
        root_delete = self.dfs(root, 0, limit)
        if root_delete:
            return None
        return root
    
    def dfs(self, node, s, limit):
        if node.left is None and node.right is None:
            return s + node.val < limit
        
        left_delete, right_delete = True, True

        if node.left:
            left_delete = self.dfs(node.left, s+node.val, limit)
        if node.right:
            right_delete = self.dfs(node.right, s+node.val, limit)
        
        if left_delete:
            node.left = None
        if right_delete:
            node.right = None
        
        return left_delete and right_delete

File: Cpp/1080_insufficient_nodes_in_root_to_leaf_paths.cpp
// https://leetcode.cn/problems/insufficient-nodes-in-root-to-leaf-paths/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sufficientSubset(TreeNode* root, int limit) {
        bool root_delete = dfs(root, 0, limit);
        if (root_delete) {
            return nullptr;
        }
        return root;
    }

    bool dfs(TreeNode *node, int s, int limit) {
        if (node->left == nullptr && node->right == nullptr) {
            return s + node->val < limit;
        }
        bool left_delete = true, right_delete = true;
        if (node->left != nullptr) {
            left_delete = dfs(node->left, s+node->val, limit);
        }
        if (node->right != nullptr) {
            right_delete = dfs(node->right, s+node->val, limit);
        }
        if (left_delete) {
            node->left = nullptr;
        }
        if (right_delete) {
            node->right = nullptr;
        }
        return left_delete && right_delete;
    }
};
